dev-tools/omdb/src/bin/omdb/db.rs

```diff
@@ -3815,3 +3887,168 @@ async fn cmd_db_reconfigurator_save(
     eprintln!("wrote {}", output_path);
     Ok(())
 }
+
+// Migrations
+
+async fn cmd_db_migrations_list(
+    datastore: &DataStore,
+    fetch_opts: &DbFetchOptions,
+    args: &MigrationsListArgs,
+) -> Result<(), anyhow::Error> {
+    use db::model::Migration;
+    use db::model::MigrationState;
+    use db::schema::migration::dsl;
+    use omicron_common::api::external::Generation;
+    use omicron_common::api::internal::nexus;
+
+    let mut state_filters = Vec::new();
+    if args.completed {
+        state_filters.push(MigrationState(nexus::MigrationState::Completed));
+    }
+    if args.failed {
+        state_filters.push(MigrationState(nexus::MigrationState::Failed));
+    }
+    if args.in_progress {
+        state_filters.push(MigrationState(nexus::MigrationState::InProgress));
+    }
+    if args.pending {
+        state_filters.push(MigrationState(nexus::MigrationState::Pending));
+    }
+
+    let mut query = dsl::migration.into_boxed();
+
+    if !fetch_opts.include_deleted {
+        query = query.filter(dsl::time_deleted.is_null());
+    }
+
+    if !state_filters.is_empty() {
+        query = query.filter(
+            dsl::source_state
+                .eq_any(state_filters.clone())
+                .or(dsl::target_state.eq_any(state_filters)),
+        );
+    }
+
+    if !args.instance_ids.is_empty() {
+        query =
+            query.filter(dsl::instance_id.eq_any(args.instance_ids.clone()));
+    }
+
+    #[derive(Tabled)]
+    #[tabled(rename_all = "SCREAMING_SNAKE_CASE")]
+    struct MigrationRow {
+        id: Uuid,
+        instance: Uuid,
+        source_vmm: Uuid,
+        source_state: MigrationState,
+        target_vmm: Uuid,
+        target_state: MigrationState,
+        created: chrono::DateTime<Utc>,
+        #[tabled(display_with = "display_option_blank")]
+        deleted: Option<chrono::DateTime<Utc>>,
+    }
+
+    #[derive(Tabled)]
+    #[tabled(rename_all = "SCREAMING_SNAKE_CASE")]
+    struct VerboseMigrationRow {
+        id: Uuid,
+        instance: Uuid,
+        src_vmm: Uuid,
+        src_state: MigrationState,
+        src_generation: Generation,
+        #[tabled(display_with = "display_option_blank")]
+        src_updated: Option<chrono::DateTime<Utc>>,
+        tgt_vmm: Uuid,
+        tgt_state: MigrationState,
+        tgt_generation: Generation,
+        #[tabled(display_with = "display_option_blank")]
+        tgt_updated: Option<chrono::DateTime<Utc>>,
+        created: chrono::DateTime<Utc>,
+        #[tabled(display_with = "display_option_blank")]
+        deleted: Option<chrono::DateTime<Utc>>,
+    }
+
+    let migrations = query
+        .limit(i64::from(u32::from(fetch_opts.fetch_limit)))
+        .order_by(dsl::time_created)
+        .select(Migration::as_select())
+        .load_async(&*datastore.pool_connection_for_tests().await?)
+        .await
+        .context("listing migrations")?;
```

Comment:
I think the full table scan is because we have no index on "time_created", so this query will need to look up all migrations to return any migrations, even less than `fetch_limit`.

Two ways of fixing this:
1. Add the index. Pros: works, no full scan. Cons: adds to cockroachdb footprint, isn't free to maintain, a little odd it's only for omdb.
2. Allow the full table scan (see: `ALLOW_FULL_TABLE_SCAN_SQL` in this file for examples). Pros: Works, minimally invasive. Cons: If we have a rack with a gargantuan number of migrations, this could cause OOM-ish issues.

Reply:
AH, yeah, that's probably the issue --- I added the order-by more recently than the last time I tested this command. I was assuming the full-table scan was necessary because we were trying to list _all_ migrations.

A third option would just be to remove the order-by clause and have `omdb` sort the result set in memory. But, because of the LIMIT clause, that has somewhat different semantics --- we will display the migrations in chronological order, but we won't necessarily have selected the _most recent_ records...which seems pretty bad. So, one of your suggestions is probably better. 

Naively, it kinda seems to me like the index is the better option...it seems like we might want to be able to order migrations by time anyway, such as if (for example) we eventually grow a UI for displaying migration history for an instance. 

And, I'd rather avoid OOM issues in "real life" racks! I notice, however, that a bunch of existing OMDB commands use the approach of allowing full-table scans instead, though, so maybe there's some reason that we would prefer not to add an index just for OMDB use? Some of those commands look like they're actually just loading the entire table, though?

Reply:
Hmm, just adding the following index still results in a full-table scan:
```sql
CREATE INDEX IF NOT EXISTS migrations_by_time_created ON omicron.public.migration (
    time_created
);
```

I wonder if it's necessary for the migration to also include `time_deleted`, or something?

Reply:
Hmm, adding `time-deleted` to the index _also_ doesn't seem to help:
```sql
CREATE INDEX IF NOT EXISTS migrations_by_time_created ON omicron.public.migration (
    time_created, time_deleted
);
```

I guess I'll just have it allow the full-table scan for now, unless someone else has any idea what I've got wrong with the index?

Reply:
I recommend looking at the `explainable` trait: https://github.com/oxidecomputer/omicron/blob/24f7cc0144a9c57875b93dfcdd85ac7a2281039e/nexus/db-queries/src/db/explain.rs#L166-L193

This emits a String from Cockroach indicating "why a full table scan occurred".

Reply:
ooh

Reply:
I don't believe the issue is with the index, but the fact that the query doesn't supply any filtering clauses on the indexed columns, e.g., `time_created`. There's nothing it can use to restrict the search here, and so this has to be a full-table scan. I believe you can add a (possibly-dummy) `time_created` to start from, and then the planner should be aware that it can use the right index.

For example:

```rust
    let migrations = query
        .limit(i64::from(u32::from(fetch_opts.fetch_limit)))
        .order_by(dsl::time_created)
        .filter(dsl::time_created.gt(chrono::DateTime::UNIX_EPOCH)) // Any time, really
        .select(Migration::as_select())
        .load_async(&*datastore.pool_connection_for_tests().await?)
        .await
        .context("listing migrations")?;
```

Reply:
For example, I ran `omicron-dev db-run` and then connected to the database. I added the same index you're using here, but a query with no filtering clauses is a full table scan, while one with a filter can use the index:

```
root@[::1]:32221/omicron> create index if not exists migration_by_creation on migration (time_created);
CREATE INDEX


Time: 111ms total (execution 111ms / network 0ms)

root@[::1]:32221/omicron> select time_created from migration;
  time_created
----------------
(0 rows)


Time: 1ms total (execution 1ms / network 0ms)

root@[::1]:32221/omicron> explain select time_created from migration;
                                     info
-------------------------------------------------------------------------------
  distribution: local
  vectorized: true

  • scan
    estimated row count: 1 (100% of the table; stats collected 5 seconds ago)
    table: migration@migration_by_creation
    spans: FULL SCAN
(7 rows)


Time: 1ms total (execution 0ms / network 0ms)

root@[::1]:32221/omicron> explain select time_created from migration where time > '1970-01-01 1:00:00 UTC';
ERROR: column "time" does not exist
SQLSTATE: 42703
root@[::1]:32221/omicron> explain select time_created from migration where time_created > '1970-01-01 1:00:00 UTC';
                                      info
--------------------------------------------------------------------------------
  distribution: local
  vectorized: true

  • scan
    estimated row count: 1 (100% of the table; stats collected 35 seconds ago)
    table: migration@migration_by_creation
    spans: [/'1970-01-01 01:00:00.000001+00:00' - ]
(7 rows)


Time: 1ms total (execution 1ms / network 0ms)
```

Reply:
here's the `EXPLAIN` output:
```
distribution: local
vectorized: true

• sort
│ estimated row count: 1
│ order: +time_created
│
└── • limit
    │ estimated row count: 1
    │ count: 500
    │
    └── • sort
        │ estimated row count: 1
        │ order: +time_created
        │
        └── • filter
            │ estimated row count: 1
            │ filter: time_deleted IS NULL
            │
            └── • scan
                  estimated row count: 1 (100% of the table; stats collected 7 minutes ago)
                  table: migration@migration_pkey
                  spans: FULL SCAN

index recommendations: 1
1. type: index creation
   SQL command: CREATE INDEX ON migration (time_deleted) STORING (instance_id, time_created, source_state, source_propolis_id, source_gen, time_source_updated, target_state, target_propolis_id, target_gen, time_target_updated);
```

and, when running with the `--include-deleted` flag, it also says:
```
distribution: local
vectorized: true

• sort
│ estimated row count: 1
│ order: +time_created
│
└── • limit
    │ estimated row count: 1
    │ count: 500
    │
    └── • sort
        │ estimated row count: 1
        │ order: +time_created
        │
        └── • scan
              estimated row count: 1 (100% of the table; stats collected 7 minutes ago)
              table: migration@migration_pkey
              spans: FULL SCAN

index recommendations: 1
1. type: index creation
   SQL command: CREATE INDEX ON migration (time_created) STORING (instance_id, time_deleted, source_state, source_propolis_id, source_gen, time_source_updated, target_state, target_propolis_id, target_gen, time_target_updated);
```

Reply:
yup, @bnaecker's suggestion of adding a no-op filter seems to fix it. Thanks Ben!